cmd/rpc/sock.go

```diff
@@ -0,0 +1,455 @@
+		// thread safety
+		r.manager.l.Lock()
+		// update the root chain info
+		r.Info = newInfo
+		// execute the callback
+		r.manager.afterRCUpdate(newInfo)
+		// release
+		r.manager.l.Unlock()
+	}
+}
+
+// Add() adds the client to the manager
+func (r *RCManager) AddSubscription(subscription *RCSubscription) {
+	// lock for thread safety
+	r.l.Lock()
+	defer r.l.Unlock()
+	// add to the map
+	r.subscriptions[subscription.chainId] = subscription
+}
+
+// RemoveSubscription() gracefully deletes a RC subscription
+func (r *RCManager) RemoveSubscription(chainId uint64) {
+	// lock for thread safety
+	r.l.Lock()
+	defer r.l.Unlock()
+	// remove from the map
+	delete(r.subscriptions, chainId)
+	// check if the chainId == a configured root chain
+	for _, rc := range r.c.RootChain {
+		// if found
+		if rc.ChainId == chainId {
+			// re-dial
+			r.NewSubscription(rc)
+			// exit
+			return
+		}
+	}
+}
+
+// Stop() stops the client
+func (r *RCSubscription) Stop(err error) {
+	// log the error
+	r.log.Errorf("WS Failed with err: %s", err.Error())
+	// close the connection
+	if err = r.conn.Close(); err != nil {
+		r.log.Error(err.Error())
+	}
+	// remove from the manager
+	r.manager.RemoveSubscription(r.chainId)
+}
+
+// SUBSCRIBER CODE BELOW (INBOUND)
+
+// RCSubscriber (Root Chain Subscriber) implements an efficient publishing service to nested chain subscribers
+type RCSubscriber struct {
+	chainId uint64          // the chain id of the publisher
+	manager *RCManager      // a reference to the manager of the ws clients
+	conn    *websocket.Conn // the underlying ws connection
+	log     lib.LoggerI     // stdout log
+}
+
+// WebSocket() upgrades a http request to a websockets connection
+func (s *Server) WebSocket(w http.ResponseWriter, r *http.Request, _ httprouter.Params) {
+	_ = w.(http.Hijacker)
+	// upgrade the connection to websockets
+	conn, err := s.rcManager.upgrader.Upgrade(w, r, nil)
+	// if an error occurred during the upgrade
+	if err != nil {
+		// write the internal server error
+		write(w, err, http.StatusInternalServerError)
+		// log the issue
+		s.logger.Error(err.Error())
+		// exit
+		return
+	}
+	// get chain id string from the parameter
+	chainIdStr := r.URL.Query().Get(chainIdParamName)
+	// get the chain id from the string
+	chainId, err := strconv.ParseUint(chainIdStr, 10, 64)
+	if err != nil {
+		http.Error(w, "invalid chain id", http.StatusBadRequest)
+		return
+	}
+	// create a new web sockets client
+	client := &RCSubscriber{
+		chainId: chainId,
+		conn:    conn,
+		manager: s.rcManager,
+		log:     s.logger,
+	}
+	// add the connection to the manager
+	s.rcManager.AddSubscriber(client)
+}
+
+// Add() adds the client to the manager
+func (r *RCManager) AddSubscriber(subscriber *RCSubscriber) {
+	// lock for thread safety
+	r.l.Lock()
+	defer r.l.Unlock()
+	// add to the map
+	r.subscribers[subscriber.chainId] = subscriber
```

Comment:
wouldn't subscribers mapped by chainID be always replaced if new ones come through? so only one subscriber per chainID could be kept, is that intentional?

Reply:
Hmmm this is a good point...

Reply:
Opened new issue #164